features/distributed/coordination/zookeeper/src/test/java/org/opennms/features/distributed/coordination/zookeeper/TestZookeeperDomainManager.java

```diff
@@ -0,0 +1,145 @@
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.IntStream;
+
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.CuratorFrameworkFactory;
+import org.apache.curator.framework.recipes.leader.CancelLeadershipException;
+import org.apache.curator.framework.recipes.leader.LeaderSelector;
+import org.apache.curator.framework.recipes.leader.LeaderSelectorListener;
+import org.apache.curator.framework.state.ConnectionState;
+import org.junit.Test;
+
+/**
+ * Tests for {@link ZookeeperDomainManager}.
+ */
+public class TestZookeeperDomainManager {
+    private final AtomicInteger connectionAttempts = new AtomicInteger(0);
+    private final AtomicInteger disconnectionAttempts = new AtomicInteger(0);
+    private final List<CompletableFuture<Void>> activeFutures = new CopyOnWriteArrayList<>();
+    private final List<CompletableFuture<Void>> standbyFutures = new CopyOnWriteArrayList<>();
+    private ZookeeperDomainManager manager;
+
+    /**
+     * Test that the registered callbacks are called when ZooKeeper informs the manager to take leadership. Also tests
+     * that only one connection attempt to ZooKeeper was initiated even though multiple registrations/deregistrations
+     * are occurring in parallel.
+     *
+     * @throws InterruptedException
+     */
+    @Test
+    public void testCallbacks() throws InterruptedException, TimeoutException, ExecutionException {
+        CuratorFramework mockClient = mock(CuratorFramework.class);
+        CuratorFrameworkFactory.Builder mockBuilder = mock(CuratorFrameworkFactory.Builder.class);
+        LeaderSelector mockLeaderSelector = mock(LeaderSelector.class);
+
+        when(mockBuilder.build()).thenReturn(mockClient);
+        doAnswer((invocation) -> connectionAttempts.incrementAndGet()).when(mockClient).start();
+        doAnswer((invocation) -> disconnectionAttempts.incrementAndGet()).when(mockClient).close();
+
+        manager = ZookeeperDomainManager.withMocks("test.domain", mockBuilder, mockLeaderSelector);
+        LeaderSelectorListener leaderSelectorListener = manager.getLeaderSelectorListener();
+
+        int numToRegister = 100;
+        IntStream.range(0, numToRegister).parallel().forEach(this::register);
+
+        // We want to wait until the connection from above actually occurred
+        await().atMost(10, TimeUnit.SECONDS).until(manager::isConnected);
+        // TODO: This has potential to false positive since we could theoretically have another thread connecting
```

Comment:
Is this a TODO, or just a note?

Reply:
More of a note. I wasn't sure how to resolve this in a clean way. Basically I need to sleep for some time before I do the assert to make sure there really wasn't any other thread attempting a connection.

The easiest thing to do would be just add a Thread.sleep(1000) before the assert and that should take care of it but I wasn't sure if that was appropriate since bare sleeps are typically discouraged in unit tests.

Reply:
I'm just going to add the 1 second sleeps in and remove those comments.